### PR TITLE
I/O chapter

### PR DESCRIPTION
**Pull request by gadjevb**
All tasks are ready for review!

### CODE REVIEW

File: src/com/clouway/io/task6/DirectoryBrowser.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import java.nio.file.*;
+import java.io.File;
+
+public class DirectoryBrowser {
+
+    public String listContent(String path) throws IOException {
+        Path dir = Paths.get(path);
+        File flagFile = new File(path);
+        if(flagFile.isFile()){
+            return flagFile.getName() + " -> is file, not a directory!";
```
Comment (mgenov) [inline src/com/clouway/io/task6/DirectoryBrowser.java]: Similar issue as before. The problem is that your method is returning String.

File: src/com/clouway/io/task4/TransferObject.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            if((offset <= 0) && (offset > size)) {
+                output.write(temp);
+            }
+            if(offset < size){
+                break;
+            }
+            count++;
+            offset--;
+        }
+        if(output != null){
```
Comment (mgenov) [inline src/com/clouway/io/task4/TransferObject.java]: finally would be better option if you are going to close streams in the TransferObject.

File: src/com/clouway/io/task4/Demo4.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+public class Demo4 {
+    public static void main(String [] args) {
+        TransferObject transfer = new TransferObject();
+        Scanner sc = new Scanner(System.in);
+        InputStream input = null;
+        OutputStream output = null;
+        String tempInput, tempOutput;
+        int offset = 0, size = 0;
+        System.out.print("Enter the input stream file: ");
```
Comment (mgenov) [inline src/com/clouway/io/task4/Demo4.java]: I see that you are using Scanner a lot of in the Demo but it's not required. 

For me something like: 

```java
TransferObject transferObject = new TransferObject();
transferObject.transfer(new FileInputStream("input.txt"), new FileOutputStream("output.txt")); 
```

is more readable then the console input. I want to see how this code is going to be used not how System.in is used as input.

File: src/com/clouway/io/task2/ConsoleInput.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package com.clouway.io.task2;
+
+import java.io.*;
+
+public class ConsoleInput {
```
Comment (mgenov) [inline src/com/clouway/io/task2/ConsoleInput.java]: ConsoleInput -> I understand it as reading from console
writeInFile -> I understand it like "writeInFile" this filename. 

The problem with naming comes from method as it does 2 things. Reading from console and writing it to the file. If you separate responsibility to 2 objects, then things will get much better. 

If you name method what it does actually, then it's name will be something like: readFromConsoleAndWriteReadToFile()
As you can see name contains "And" which is a smell that method is doing 2 things at a time which is a programming SMELL and this code needs refactoring. 

As example, you can take a look this code:
```java
ConsoleReader reader = new ConsoleReader();
String content = reader.readTillReach("."); //as you can see I'm able to specify and separator ...

TextFile textFile = new TextFile("result.txt");
textFile.write(content);
```

File: src/com/clouway/io/task3/Reverse.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package com.clouway.io.task3;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Reverse {
```
Comment (mgenov) [inline src/com/clouway/io/task3/Reverse.java]: Reverse is not a good name for a class cause it's a verb. 

FileObject, TextFile, TextStream or something similar are good candidates.

File: src/com/clouway/io/task3/Demo3.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.clouway.io.task3;
+
+import java.io.IOException;
+
+public class Demo3 {
```
Comment (mgenov) [inline src/com/clouway/io/task3/Demo3.java]: You don't have to make Demo1, Demo2, Demo3 as they all are in separate packages, so just Demo is more then enough.

**Comment by gadjevb** [comment]
Initially they were named Demo, but when I tried to run and test the code I was seeing only Demo's. I didn't know which one was the correct, so I gave them names in order to distinct them.

**Comment by mgenov** [comment]
If so,  you can use name that is topic related. 

SerializationDemo, ConsoleReaderDemo and etc.

**Comment by gadjevb** [comment]
They are renamed now. The first and the second tasks have very similar names, I will think about better ones.

File: src/com/clouway/io/task3/TextFile.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package com.clouway.io.task3;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class TextFile {
+
+    public String reverseFile(String fileName) throws IOException {
```
Comment (mgenov) [inline src/com/clouway/io/task3/TextFile.java]: The signature of this method is bad cause TextFile represents single file but reverseFile could reverse many files which is misleading. 
```java
new TextFile("names.txt").reverse(); 
// or
new TextFilesApi().reverseFile("names.txt"); 
```

**Comment by mgenov** [comment]
LGTM !

**Pull request closed without merging**